source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp

```diff
@@ -128,180 +127,119 @@ MPxNode::SchedulingType MaterialXNode::schedulingType() const
-				{
-                    float2& value = dataHandle.asFloat2();
-                    valueElement->setValue(MaterialX::Color2(value[0], value[1]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Color3>::TYPE)
-				{
-					float3& value = dataHandle.asFloat3();
-					valueElement->setValue(MaterialX::Color3(value[0], value[1], value[2]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Color4>::TYPE)
-				{
-					MFloatVector& value = dataHandle.asFloatVector();
-                    valueElement->setValue(MaterialX::Color4(value[0], value[1], value[2], value[3]));
-				}
-				else if (type == MaterialX::TypedValue<float>::TYPE)
-				{
-					float& value = dataHandle.asFloat();
-					valueElement->setValue(value);
-				}
-			}
-		}
-	}
-	return true;
+    return true;
 }
 
-void MaterialXNode::setAttributeValue(MObject &materialXObject, MObject &attr, float* values, unsigned int size, MDGModifier& mdgModifier)
+bool MaterialXNode::setInternalValue(const MPlug& plug, const MDataHandle& dataHandle)
 {
-	MPlug plug(materialXObject, attr);
-	if (size == 1)
-	{
-		mdgModifier.newPlugValueDouble(plug, values[0]);
-	}
-	else
-	{
-		for(unsigned int i=0; i<size; i++)
-		{
-			MPlug indexPlug = plug.child(i);
-			mdgModifier.newPlugValueDouble(indexPlug, values[i]);
-		}
-	}
-}
+    auto displayError = [this](const char* error)
+    {
+        MString message("Failed to create shader for ");
+        message += typeName();
+        message += ' ';
+        message += name();
+        message += ": ";
+        message += error;
+        MGlobal::displayError(message);
+    };
+
+    auto loadDocument = [this]()
+    {
+        return MaterialXMaya::loadDocument(
+            _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
+        );
+    };
 
-void MaterialXNode::createAttributesFromDocument(MDGModifier& mdgModifier)
-{
-    MaterialX::DocumentPtr document;
-    if (!materialXData || !(document = materialXData->getDocument()))
+    if (plug == DOCUMENT_ATTRIBUTE)
+    {
+        const MString& value = dataHandle.asString();
+        if (_documentFilePath == value)
+        {
+            return true;
+        }
+
+        _documentFilePath = value;
+        _materialXData.reset();
+
+        try
+        {
+            _materialXData.reset( new MaterialXData(
+                loadDocument(),
+                _xmlElementPath.asChar(),
+                Plugin::instance().getLibrarySearchPath()
+            ));
+        }
+        catch (std::exception& e)
+        {
+            displayError(e.what());
+        }
+    }
+    else if (plug == ELEMENT_ATTRIBUTE)
+    {
+        const MString& value = dataHandle.asString();
+        if (_xmlElementPath == value)
+        {
+            return true;
+        }
+
+        _xmlElementPath = value;
+        mx::DocumentPtr document;
+
+        if (_materialXData)
+        {
+            document = _materialXData->getDocument();
```

Comment:
If a valid document existed, reuse it. The node should probably reference the document directly, otherwise it would get destroyed if the new MaterialXData fails to be created.